### PR TITLE
[bitnami/deepspeed] feat: :tada: Add chart

### PR DESCRIPTION
**Pull request by javsalgar**
### Description of the change

This PR adds bitnami/deepspeed to the catalog. It has the following features:

 -  Bitnami standards
 -  Deploy as Job or Deployment
 -  Custom configuration and projects via configmaps or git
 -  SSH connection via keys instead of passwordless

### Benefits

New product to the chart

### Possible drawbacks

n/a

### Applicable issues

- fixes #

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: bitnami/deepspeed/templates/client/hostfile-configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  {{- if .Values.commonAnnotations }}
+  annotations: {{- include "common.tplvalues.render" (dict "value" .Values.commonAnnotations "context" $) | nindent 4 }}
+  {{- end }}
+data:
+  hostfile: |
+    {{- if .Values.overrideHostFile }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.config.overrideHostFile "context" $) | nindent 4 }}
+    {{- else }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.config.defaultHostFile "context" $) | nindent 4 }}
+    {{- end }}
```
Comment (fmulero) [inline bitnami/deepspeed/templates/client/hostfile-configmap.yaml:27]: Not sure why and it isn't a big problem, when template is rendered the `hostfile` starts with a blank line

**Review by fmulero — COMMENTED** [review]
Great work @javsalgar! I like the new version identifier in the helpers !

I left some comments, Could you review them?

File: bitnami/deepspeed/templates/client/hostfile-configmap.yaml (hunk 1 above)
Comment (fmulero) [reply bitnami/deepspeed/templates/client/hostfile-configmap.yaml]: You right, after installation CM shows the right content, helm template shows the blank line at the beginning. We can close this conversation, it shouldn't be a problem.

**Review by fmulero — DISMISSED** [review]
LGTM

File: bitnami/deepspeed/templates/client/networkpolicy.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        - port: 53
+          protocol: TCP
+    # Allow outbound connections to other cluster pods
+    - ports:
+        - port: {{ .Values.worker.containerPorts.ssh }}
+        - port: {{ .Values.worker.externalAccess.service.ports.ssh }}
+      to:
+        - podSelector:
+            matchLabels: {{- include "common.labels.standard" . | nindent 14 }}
+              {{- if .Values.podLabels }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/client/networkpolicy.yaml:43]: Value `podLabels` is missing in the values.yaml

File: bitnami/deepspeed/templates/client/client-dep-job.yaml [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    {{- end }}
+    {{- if .Values.commonLabels }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.commonLabels "context" $) | nindent 4 }}
+    {{- end }}
+spec:
+  selector:
+    matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
+      app.kubernetes.io/part-of: deepspeed
+      app.kubernetes.io/component: client
+  {{- if .Values.useJob }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/client/client-dep-job.yaml:33]: Shouldn't this be `.Values.client.useJob`?

File: bitnami/deepspeed/templates/worker/worker-service.yaml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  {{- end }}
+  {{- if or (eq $root.Values.worker.externalAccess.service.type "LoadBalancer") (eq $root.Values.worker.externalAccess.service.type "NodePort") }}
+  externalTrafficPolicy: {{ $root.Values.worker.externalAccess.service.externalTrafficPolicy | quote }}
+  {{- end }}
+  publishNotReadyAddresses: {{ $root.Values.worker.externalAccess.service.publishNotReadyAddresses }}
+  {{- if $root.Values.worker.externalAccess.service.sessionAffinity }}
+  sessionAffinity: {{ $root.Values.worker.externalAccess.service.sessionAffinity }}
+  {{- end }}
+  {{- if $root.Values.worker.externalAccess.service.sessionAffinityConfig }}
+  sessionAffinityConfig: {{- include "common.tplvalues.render" (dict "value" $root.Values.service.sessionAffinityConfig "context" $root) | nindent 4 }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/worker/worker-service.yaml:58]:
```suggestion
  sessionAffinityConfig: {{- include "common.tplvalues.render" (dict "value" $root.Values.worker.externalAccess.service.sessionAffinityConfig "context" $root) | nindent 4 }}
```

File: bitnami/deepspeed/templates/client/hostfile-configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    {{- include "common.tplvalues.render" (dict "value" .Values.commonLabels "context" $) | nindent 4 }}
+    {{- end }}
+    app.kubernetes.io/part-of: deepspeed
+    app.kubernetes.io/component: client
+  {{- if .Values.commonAnnotations }}
+  annotations: {{- include "common.tplvalues.render" (dict "value" .Values.commonAnnotations "context" $) | nindent 4 }}
+  {{- end }}
+data:
+  hostfile: |
+    {{- if .Values.overrideHostFile }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/client/hostfile-configmap.yaml:23]:
```suggestion
    {{- if .Values.config.overrideHostFile }}
```

File: bitnami/deepspeed/templates/_helpers.tpl [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+  imagePullPolicy: {{ default "" .context.Values.volumePermissions.image.pullPolicy | quote }}
+  command:
+    - /bin/bash
+  args:
+    - -ec
+    - |
+      #!/bin/bash
+      mkdir -p {{ $block.persistence.mountPath }}
+      chown {{ $block.containerSecurityContext.runAsUser }}:{{ $block.podSecurityContext.fsGroup }} {{ .context.Values.persistence.mountPath }}
+      find {{ $block.client.persistence.mountPath }} -mindepth 1 -maxdepth 1 -not -name ".snapshot" -not -name "lost+found" | xargs chown -R {{ $block.containerSecurityContext.runAsUser }}:{{ $block.podSecurityContext.fsGroup }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/_helpers.tpl:265]:
```suggestion
      mkdir -p {{ $block.persistence.mountPath }}
      chown {{ $block.containerSecurityContext.runAsUser }}:{{ $block.podSecurityContext.fsGroup }} {{ $block.persistence.mountPath }}
      find {{ $block.persistence.mountPath }} -mindepth 1 -maxdepth 1 -not -name ".snapshot" -not -name "lost+found" | xargs chown -R {{ $block.containerSecurityContext.runAsUser }}:{{ $block.podSecurityContext.fsGroup }}
```

File: bitnami/deepspeed/templates/worker/worker-statefulset.yaml [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+        {{- end }}
+      spec:
+        accessModes:
+        {{- range .Values.worker.persistence.accessModes }}
+          - {{ . | quote }}
+        {{- end }}
+        resources:
+          requests:
+            storage: {{ .Values.worker.persistence.size | quote }}
+        {{- if .Values.worker.persistence.selector }}
```
[the remark this answers is not in the document]
Comment (migruiz4) [reply bitnami/deepspeed/templates/worker/worker-statefulset.yaml:272]: Conversation marked as resolved, but value is still missing

File: bitnami/deepspeed/templates/worker/worker-statefulset.yaml [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+kind: StatefulSet
+metadata:
+  name: {{ include "deepspeed.v0.worker.fullname" . }}
+  namespace: {{ include "common.names.namespace" . | quote }}
+  labels: {{- include "common.labels.standard" . | nindent 4 }}
+    app.kubernetes.io/part-of: deepspeed
+    app.kubernetes.io/component: worker
+    {{- if .Values.labels }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.labels "context" $) | nindent 4 }}
+    {{- end }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/worker/worker-statefulset.yaml:17]: `.Values.labels` is missing in the values.yaml

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Please take a look at my comments

File: bitnami/deepspeed/templates/worker/worker-service.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  {{- end }}
+  {{- if or (eq $root.Values.worker.externalAccess.service.type "LoadBalancer") (eq $root.Values.worker.externalAccess.service.type "NodePort") }}
+  externalTrafficPolicy: {{ $root.Values.worker.externalAccess.service.externalTrafficPolicy | quote }}
+  {{- end }}
+  publishNotReadyAddresses: {{ $root.Values.worker.externalAccess.service.publishNotReadyAddresses }}
+  {{- if $root.Values.worker.externalAccess.service.sessionAffinity }}
+  sessionAffinity: {{ $root.Values.worker.externalAccess.service.sessionAffinity }}
+  {{- end }}
+  {{- if $root.Values.worker.externalAccess.service.sessionAffinityConfig }}
+  sessionAffinityConfig: {{- include "common.tplvalues.render" (dict "value" $root.Values.externalAccess.service.sessionAffinityConfig "context" $root) | nindent 4 }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/worker/worker-service.yaml:58]: My bad, I was missing the `worker` on my previous suggestion
```suggestion
  sessionAffinityConfig: {{- include "common.tplvalues.render" (dict "value" $root.Values.worker.externalAccess.service.sessionAffinityConfig "context" $root) | nindent 4 }}
```

File: bitnami/deepspeed/templates/client/client-dep-job.yaml [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+    {{- end }}
+    {{- if .Values.commonAnnotations }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.commonAnnotations "context" $) | nindent 8 }}
+    {{- end }}
+  labels: {{- include "common.labels.standard" . | nindent 4 }}
+    app.kubernetes.io/part-of: deepspeed
+    app.kubernetes.io/component: client
+    {{- if .Values.client.labels }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.client.labels "context" $) | nindent 4 }}
+    {{- end }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/client/client-dep-job.yaml:24]: `client.labels` missing in values.yaml

File: bitnami/deepspeed/templates/worker/worker-statefulset.yaml [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ include "deepspeed.v0.worker.fullname" . }}
+  namespace: {{ include "common.names.namespace" . | quote }}
+  labels: {{- include "common.labels.standard" . | nindent 4 }}
+    app.kubernetes.io/part-of: deepspeed
+    app.kubernetes.io/component: worker
+    {{- if .Values.worker.labels }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.labels "context" $) | nindent 4 }}
```
Comment (migruiz4) [inline bitnami/deepspeed/templates/worker/worker-statefulset.yaml:16]:
```suggestion
    {{- include "common.tplvalues.render" (dict "value" .Values.worker.labels "context" $) | nindent 4 }}
```

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Sorry, a few more comments

**Review by migruiz4 — APPROVED** [review]
Looks good to me!

**Pull request merged**